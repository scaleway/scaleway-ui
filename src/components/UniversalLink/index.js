import { useTheme } from '@emotion/react'
import PropTypes from 'prop-types'
import React from 'react'
import Box from '../Box'

const ABSOLUTE_LINK_REGEXP = /^https?:\/\//
const TEL_LINK_REGEXP = /^tel:/
const MAILTO_LINK_REGEXP = /^mailto:/

const needNativeLink = url => {
  if (!url) return false
  const isAbsolute = ABSOLUTE_LINK_REGEXP.test(url)
  const isTelLink = TEL_LINK_REGEXP.test(url)
  const isMailToLink = MAILTO_LINK_REGEXP.test(url)
  const isAnchor = url[0] === '#'

  return isAbsolute || isTelLink || isMailToLink || isAnchor
}

const UniversalLink = ({
  children,
  target,
  rel: propsRel,
  to,
  as: propsAs,
  href: propsHref,
  ...props
}) => {
  const { linkComponent = 'a' } = useTheme()
  const isBlank = target === '_blank'
  const rel = propsRel || (isBlank ? 'noopener noreferrer' : undefined)
  const href = to || propsHref
  const asValue = propsAs || (needNativeLink(href) ? 'a' : linkComponent)
  const parameter = asValue === 'a' ? 'href' : 'to'

  return (
    <Box
      {...props}
      target={target}
      as={asValue}
      {...{ [parameter]: href }}
      rel={rel}
    >
      {children}
    </Box>
  )
}

UniversalLink.propTypes = {
  as: PropTypes.node,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
  href: PropTypes.string,
  rel: PropTypes.string,
  target: PropTypes.string,
  to: PropTypes.string,
}

UniversalLink.defaultProps = {
  as: null,
  href: null,
  rel: undefined,
  target: undefined,
  to: null,
}

export default UniversalLink
