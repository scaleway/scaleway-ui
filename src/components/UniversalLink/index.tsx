import { useTheme } from '@emotion/react'
import PropTypes from 'prop-types'
import React, { ComponentProps, ElementType, ReactNode, Ref } from 'react'
import Box, { XStyledProps } from '../Box'

const ABSOLUTE_LINK_REGEXP = /^https?:\/\//
const TEL_LINK_REGEXP = /^tel:/
const MAILTO_LINK_REGEXP = /^mailto:/

const needNativeLink = (url?: string) => {
  if (!url) return false
  const isAbsolute = ABSOLUTE_LINK_REGEXP.test(url)
  const isTelLink = TEL_LINK_REGEXP.test(url)
  const isMailToLink = MAILTO_LINK_REGEXP.test(url)
  const isAnchor = url[0] === '#'

  return isAbsolute || isTelLink || isMailToLink || isAnchor
}

export type UniversalLinkProps = {
  children: ReactNode
  to?: string
} & ComponentProps<'a'> & {
    ref?: Ref<Element>
  } & XStyledProps

const UniversalLink = ({
  children,
  target,
  rel: propsRel,
  to,
  as: propsAs,
  href: propsHref,
  ...props
}: UniversalLinkProps): JSX.Element => {
  const { linkComponent = 'a' } = useTheme()
  const isBlank = target === '_blank'
  const rel = propsRel || (isBlank ? 'noopener noreferrer' : undefined)
  const href = to || propsHref
  const asValue =
    propsAs ||
    (needNativeLink(href) ? 'a' : (linkComponent as ElementType<unknown>))
  const parameter = asValue === 'a' ? 'href' : 'to'

  return (
    <Box
      {...props}
      target={target}
      as={asValue}
      {...{ [parameter]: href }}
      rel={rel}
    >
      {children}
    </Box>
  )
}

UniversalLink.propTypes = {
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  href: PropTypes.string,
  rel: PropTypes.string,
  target: PropTypes.string,
  to: PropTypes.string,
}

export default UniversalLink
