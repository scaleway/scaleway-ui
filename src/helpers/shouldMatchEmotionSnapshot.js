/* eslint-disable import/no-extraneous-dependencies */
import createCache from '@emotion/cache'
import { createSerializer } from '@emotion/jest'
import { CacheProvider, ThemeProvider } from '@emotion/react'
import { render } from '@testing-library/react'
import React from 'react'
import theme from '../theme'

expect.addSnapshotSerializer(createSerializer())

const emotionCache = createCache({
  key: 'cache',
})
emotionCache.compat = true

export default async (component, { transform } = {}) => {
  const node = render(
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </CacheProvider>,
  )
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
}
