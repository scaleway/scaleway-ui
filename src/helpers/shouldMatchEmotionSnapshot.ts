/* eslint-disable import/no-extraneous-dependencies */
import { createSerializer } from '@emotion/jest'
import { RenderOptions, render } from '@testing-library/react'
import { ReactNode } from 'react'
import defaultTheme from '../theme'
import renderWithTheme from './renderWithTheme'

// use only class hash (generated from css style content)
expect.addSnapshotSerializer(
  createSerializer({ classNameReplacer: className => className }),
)

interface Options {
  options?: RenderOptions
  theme?: typeof defaultTheme
  transform?: (node: ReturnType<typeof render>) => Promise<void> | void
}

export default async (
  component: ReactNode,
  { transform, options, theme }: Options = {},
): Promise<void> => {
  const node = renderWithTheme(component, options, theme)
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
  node.unmount()
}
