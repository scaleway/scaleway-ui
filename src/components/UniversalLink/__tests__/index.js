import React from 'react'
import { BrowserRouter, Link } from 'react-router-dom'
import UniversalLink from '..'
import shouldMatchEmotionSnapshot from '../../../helpers/shouldMatchEmotionSnapshot'

describe('UniversalLink', () => {
  test.each`
    test                                        | href                               | to                          | target       | as
    ${'renders correctly with defaults'}        | ${undefined}                       | ${undefined}                | ${undefined} | ${undefined}
    ${'renders correctly with href'}            | ${'/'}                             | ${undefined}                | ${undefined} | ${undefined}
    ${'renders correctly with to prop'}         | ${undefined}                       | ${'/'}                      | ${undefined} | ${undefined}
    ${'renders correctly with an absolute utl'} | ${undefined}                       | ${'https://google.com'}     | ${'_blank'}  | ${undefined}
    ${'renders correctly with a target blank'}  | ${undefined}                       | ${'/'}                      | ${'_blank'}  | ${undefined}
    ${'renders correctly with as Link'}         | ${undefined}                       | ${'/test-react-router-dom'} | ${undefined} | ${Link}
    ${'renders correctly with as a'}            | ${'https://react.ui.scaleway.com'} | ${undefined}                | ${undefined} | ${'a'}
    ${'renders correctly with a tel link'}      | ${undefined}                       | ${'tel:+33666666'}          | ${'_blank'}  | ${undefined}
    ${'renders correctly with a mailto link'}   | ${undefined}                       | ${'mailto:test@test.com'}   | ${'_blank'}  | ${undefined}
  `('$test', ({ href, to, target, as }) => {
    shouldMatchEmotionSnapshot(
      <BrowserRouter>
        <UniversalLink href={href} to={to} target={target} as={as}>
          This is a link
        </UniversalLink>
      </BrowserRouter>,
    )
  })

  test('renders correctly with modify Link on theme from react-router-dom', () => {
    const theme = {
      linkComponent: Link,
    }
    shouldMatchEmotionSnapshot(
      <BrowserRouter>
        <UniversalLink to="/test-react-router-dom">
          This is a link
        </UniversalLink>
      </BrowserRouter>,
      { theme },
    )
  })
})
