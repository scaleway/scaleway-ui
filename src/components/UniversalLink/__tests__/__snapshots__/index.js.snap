// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UniversalLink renders correctly with a taregt blank 1`] = `
<DocumentFragment>
  <a
    class="sui-box emotion-0"
    href="/"
    rel="noopener noreferrer"
    target="_blank"
    to="/"
  >
    This is a link
  </a>
</DocumentFragment>
`;

exports[`UniversalLink renders correctly with an absolute utl 1`] = `
<DocumentFragment>
  <a
    class="sui-box emotion-0"
    href="https://google.com"
    rel="noopener noreferrer"
    target="_blank"
    to="https://google.com"
  >
    This is a link
  </a>
</DocumentFragment>
`;

exports[`UniversalLink renders correctly with defaults 1`] = `
<DocumentFragment>
  <a
    class="sui-box emotion-0"
  >
    This is a link
  </a>
</DocumentFragment>
`;

exports[`UniversalLink renders correctly with href 1`] = `
<DocumentFragment>
  <a
    class="sui-box emotion-0"
    href="/"
  >
    This is a link
  </a>
</DocumentFragment>
`;

exports[`UniversalLink renders correctly with to prop 1`] = `
<DocumentFragment>
  <a
    class="sui-box emotion-0"
    href="/"
    to="/"
  >
    This is a link
  </a>
</DocumentFragment>
`;
